File: test/repo/repo_cli.c
/*
 *  Copyright (C) 2014, 2015, 2016 Michael Hofmann
 *  
 *  This file is part of the Simulation Component and Data Coupling (SCDC) library.
 *  
 *  The SCDC library is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  The SCDC library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser Public License for more details.
 *  
 *  You should have received a copy of the GNU Lesser Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <signal.h>
#include <getopt.h>

#include "z_pack.h"

#include "scdc.h"

#include "repoH.h"


const char *prefix = "repo_cli: ";


void sighandler(int sig)
{
  printf("sighandler: sig: %d\n", sig);

  fclose(stdin);
}


scdcint_t manual_input_next(scdc_dataset_input_t *input)
{
  scdcint_t i = (intptr_t) input->data;

  printf("INPUT #%" scdcint_fmt " + <Enter>: ", i);

  ((char *) SCDC_DATASET_INOUT_BUF_PTR(input))[0] = 0;

  fgets(SCDC_DATASET_INOUT_BUF_PTR(input), SCDC_DATASET_INOUT_BUF_SIZE(input), stdin);

  SCDC_DATASET_INOUT_BUF_CURRENT(input) = strlen(SCDC_DATASET_INOUT_BUF_PTR(input));

  if (!feof(stdin)) --SCDC_DATASET_INOUT_BUF_CURRENT(input);

  input->total_size += SCDC_DATASET_INOUT_BUF_CURRENT(input);

  if (feof(stdin) || SCDC_DATASET_INOUT_BUF_CURRENT(input) <= 0) input->next = NULL;

  input->data = (void *) (intptr_t) (i + 1);

  return SCDC_SUCCESS;
}


scdcint_t manual_output_next(scdc_dataset_output_t *output)
{
  scdcint_t i = (intptr_t) output->data;

  if (strcmp(output->format, "text") == 0 || strcmp(output->format, "fslist") == 0)
    printf("OUTPUT #%" scdcint_fmt ": '%.*s'\n", i, (int) SCDC_DATASET_INOUT_BUF_CURRENT(output), (char *) SCDC_DATASET_INOUT_BUF_PTR(output));
  else
    printf("OUTPUT #%" scdcint_fmt": %" scdcint_fmt " bytes\n", i, SCDC_DATASET_INOUT_BUF_CURRENT(output));

  output->data = (void *) (intptr_t) (i + 1);

  return SCDC_SUCCESS;
}



#define DEFAULT_INPUT_BUF_SIZE   256
#define DEFAULT_OUTPUT_BUF_SIZE  256


static char default_input_buf[DEFAULT_INPUT_BUF_SIZE], default_output_buf[DEFAULT_OUTPUT_BUF_SIZE];


void prepare_dataset_cmd(const char *cmdline, char *cmd, scdc_dataset_input_t **input, int *input_type, scdc_dataset_output_t **output, int *output_type)
{
  char *input_pos, *output_pos, *name_pos, *type_pos;


  input_pos = strchr(cmdline, '<');
  output_pos = strchr(cmdline, '>');

  if (input_pos) { *input_pos = '\0'; ++input_pos; }
  if (output_pos) { *output_pos = '\0'; ++output_pos; }

  *input_type = 0;

  if (input_pos)
  {
    *input_type = 1;

    name_pos = strchr(input_pos, ':');

    if (name_pos) { *name_pos = '\0'; ++name_pos; type_pos = input_pos; }
    else { name_pos = input_pos; type_pos = NULL; }

    if (!type_pos)
    {
      if (strlen(name_pos) > 0)
      {
        if (z_fs_is_directory(name_pos)) type_pos = "fs";
        else if (z_fs_is_file(name_pos)) type_pos = "file";
        else type_pos = "text";

      } else type_pos = "manual";
    }

    if (strcmp(type_pos, "fs") == 0)
    {
      printf("input fs: '%s'\n", name_pos);
      *input = scdc_dataset_input_create(*input, "fs", ".", name_pos);

    } else if (strcmp(type_pos, "file") == 0)
    {
      printf("input file: '%s'\n", name_pos);
      *input = scdc_dataset_input_create(*input, "file", name_pos);

    } else if (strcmp(type_pos, "stream") == 0)
    {
      printf("input stream: '%s'\n", name_pos);
      *input = scdc_dataset_input_create(*input, "stream", stdin);

    } else if (strcmp(type_pos, "text") == 0)
    {
      printf("input text: '%s'\n", name_pos);
      *input = scdc_dataset_input_create(*input, "buffer", name_pos, strlen(name_pos));

      strcpy((*input)->format, "text");

    } else
    {
      printf("input manual:\n");

      scdc_dataset_input_unset(*input);

      strcpy((*input)->format, "text");

      SCDC_DATASET_INOUT_BUF_PTR(*input) = default_input_buf;
      SCDC_DATASET_INOUT_BUF_SIZE(*input) = DEFAULT_INPUT_BUF_SIZE;

      SCDC_DATASET_INOUT_BUF_CURRENT(*input) = 0;
      (*input)->total_size = 0;
      (*input)->total_size_given = SCDC_DATASET_INOUT_TOTAL_SIZE_GIVEN_AT_LEAST;

      (*input)->next = manual_input_next;

      (*input)->data = (void *) 0;

      *input_type = -1;
    }

  } else
  {
    printf("input null:\n");
    *input = NULL;
  }


  *output_type = 0;

  if (output_pos)
  {
    *output_type = 1;

    name_pos = strchr(output_pos, ':');

    if (name_pos) { *name_pos = '\0'; ++name_pos; type_pos = output_pos; }
    else { name_pos = output_pos; type_pos = NULL; }

    if (!type_pos) type_pos = "redirect";

    if (strcmp(type_pos, "fs") == 0)
    {
      printf("output fs: '%s'\n", name_pos);
      *output = scdc_dataset_output_create(*output, "fs", name_pos);

    } else if (strcmp(type_pos, "file") == 0)
    {
      printf("output file: '%s'\n", name_pos);
      *output = scdc_dataset_output_create(*output, "file", name_pos);

    } else if (strcmp(type_pos, "stream") == 0)
    {
      printf("output stream: '%s'\n", name_pos);
      *output = scdc_dataset_output_create(*output, "stream", stdout);

    } else if (strcmp(type_pos, "consume") == 0)
    {
      printf("output consume:\n");
      *output = scdc_dataset_output_create(*output, "consume");

    } else if (strcmp(type_pos, "redirect") == 0)
    {
      printf("output redirect:\n");
      *output = scdc_dataset_output_create(*output, "alloc");
      *output_type = 2;

    } else
    {
      printf("output manual:\n");

      scdc_dataset_output_unset(*output);

      SCDC_DATASET_INOUT_BUF_PTR(*output) = default_output_buf;
      SCDC_DATASET_INOUT_BUF_SIZE(*output) = DEFAULT_OUTPUT_BUF_SIZE;

      (*output)->next = manual_output_next;

      (*output)->data = (void *) 0;

      *output_type = -1;
    }

  } else
  {
    printf("output loop manual:\n");

    scdc_dataset_output_unset(*output);

    SCDC_DATASET_INOUT_BUF_PTR(*output) = default_output_buf;
    SCDC_DATASET_INOUT_BUF_SIZE(*output) = DEFAULT_OUTPUT_BUF_SIZE;
  }

  strcpy(cmd, cmdline);
}


void unprepare_dataset_cmd(scdc_dataset_input_t *input, int input_type, scdc_dataset_output_t *output, int output_type)
{
  scdcint_t i;


  printf("cmd input: ");
  scdc_dataset_input_print(input);
  printf("\n");

  if (input_type > 0)
  {
    scdc_dataset_input_destroy(input);

  } else if (input)
  {
    /* manual */
  }

  printf("cmd output: ");
  scdc_dataset_output_print(output);
  printf("\n");

  if (output_type > 0)
  {
    if (output_type == 2) scdc_dataset_output_redirect(output, "stream", stdout);

    scdc_dataset_output_destroy(output);

  } else if (output)
  {
    /* manual */
    if (output_type == 0)
    {
      i = 0;
      do {
        void *data = output->data;
        output->data = (void *) (intptr_t) i;

        manual_output_next(output);
#if 0
        break;
#endif

        i = (intptr_t) output->data;
        output->data = data;

        if (!output->next) break;

#if 1
        printf("Press <Enter>");
        getchar();
#endif

        output->next(output);

      } while (1);
    }
  }
}


void run_cli(scdc_dataset_t dataset, char *cmdline)
{
  int cmdline_given;
  size_t cmdline_size = 256;
  ssize_t n;

#define CMD_SIZE  256
  char cmd[CMD_SIZE];
  scdc_dataset_input_t _input, *input, _output, *output;
  int input_type, output_type;


  cmdline_given = (cmdline && strlen(cmdline) > 0);

  if (!cmdline_given) cmdline = malloc(cmdline_size);

  while (1)
  {
    if (!cmdline_given)
    {
      printf("Enter command for dataset '%p': ", dataset);
      n = getline(&cmdline, &cmdline_size, stdin);

    } else n = strlen(cmdline);

    if (n > 0 && cmdline[n - 1] == '\n') 
    {
      cmdline[n - 1] = '\0';
      --n;
    }

    if (n < 0 || feof(stdin) || strcmp(cmdline, "quit") == 0) break;

    if (n == 0) continue;

    input = &_input;
    output = &_output;

    prepare_dataset_cmd(cmdline, cmd, &input, &input_type, &output, &output_type);

    scdcint_t ret = scdc_dataset_cmd(dataset, cmd, input, output);

    unprepare_dataset_cmd(input, input_type, output, output_type);

    printf("%scommand '%s' %s!\n", prefix, cmd, ((ret == SCDC_SUCCESS)?"success":"failed"));

    if (cmdline_given) break;
  }

  if (!cmdline_given) free(cmdline);
}


int main(int argc, char *argv[])
{
#define DATAPROVS_MAX  16
  int ndataprovs = 0;
  scdc_dataprov_t dataprovs[DATAPROVS_MAX];

  scdc_dataset_t dataset;

#define OPTARG_MAX  256
  char scheme[OPTARG_MAX], authority[OPTARG_MAX], open[OPTARG_MAX], cmdline[OPTARG_MAX];
  char path[OPTARG_MAX], *e;


  printf("start repository client\n");

  signal(SIGABRT, &sighandler);
  signal(SIGTERM, &sighandler);
  signal(SIGINT, &sighandler);

#define SETUP  0

#if SETUP == 0
  strcpy(scheme, "scdc");
  strcpy(authority, "");
  strcpy(open, "repoA");
  strcpy(cmdline, "ls>stream:");
#elif SETUP == 1
  strcpy(scheme, "scdc+tcp");
  strcpy(authority, "localhost");
  strcpy(open, "repoA");
  strcpy(cmdline, "");
#elif SETUP == 2
  strcpy(scheme, "scdc");
  strcpy(authority, "");
  strcpy(open, "repoB/xxx");
  strcpy(cmdline, "");
#elif SETUP == 3
  strcpy(scheme, "scdc+tcp");
  strcpy(authority, "localhost");
  strcpy(open, "repoD/");
  strcpy(cmdline, "");
#elif SETUP == 4
  strcpy(scheme, "scdc+tcp");
  strcpy(authority, "localhost");
  strcpy(open, "www/");
  strcpy(cmdline, "");
#elif SETUP == 5
  strcpy(scheme, "scdc+tcp");
  strcpy(authority, "patty");
  strcpy(open, "www");
  strcpy(cmdline, "");
#elif SETUP == 6
  strcpy(scheme, "scdc");
  strcpy(authority, "");
  strcpy(open, "repoD");
  strcpy(cmdline, "");
#elif SETUP == 7
  strcpy(scheme, "scdc+tcp");
  strcpy(authority, "localhost");
  strcpy(open, "repoH");
  strcpy(cmdline, "");
#elif SETUP == 8
  strcpy(scheme, "scdc");
  strcpy(authority, "");
  strcpy(open, "repoC");
  strcpy(cmdline, "");
#elif SETUP == 9
  strcpy(scheme, "scdc");
  strcpy(authority, "");
  strcpy(open, "repoJ");
  strcpy(cmdline, "");
#endif

#if 0
  strcpy(scheme, "scdc+mpi");
  strcpy(authority, "publ:TESTNAME:0");
  strcpy(open, "repoA/");
  strcpy(cmdline, "");
#endif

#if 1
  strcpy(scheme, "");
  strcpy(authority, "");
  strcpy(open, "");
/*  strcpy(cmdline, "scdc+tcp://localhost/repoR/relA ls");*/
  strcpy(cmdline, "scdc+uds:///repoA ls");
#endif

/*  scdc_init(SCDC_INIT_DEFAULT ":log_filepath", "scdclog");*/
  scdc_init(SCDC_INIT_DEFAULT);

#if 1
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoA", "fs_access", (e = getenv("MERGE_SCDC_REPO_PATH"), sprintf(path, "%s%s", (e?e:""), "A"), path));
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoB", "fs_access", (e = getenv("MERGE_SCDC_REPO_PATH"), sprintf(path, "%s%s", (e?e:""), "B"), path));
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoC", "gen");
/*  dataprovs[ndataprovs++] = scdc_dataprov_open("repoD", "mysql", getenv("MERGE_SCDC_MYSQL_CREDENTIALS"));*/
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoH", "hook:id", &repoH, 0x2501);
/*  dataprovs[ndataprovs++] = scdc_dataprov_open("repoJ", "jobrun", "uname -a; sleep 1", (e = getenv("MERGE_SCDC_REPO_PATH"), sprintf(path, "%s%s", (e?e:""), "J"), path));
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoR", "jobrun_relay:relay:relay", "relA scdc:///repoA", "relB scdc:///repoB");*/
  dataprovs[ndataprovs++] = scdc_dataprov_open("repoS", "fs_store", (e = getenv("MERGE_SCDC_REPO_PATH"), sprintf(path, "%s%s", (e?e:""), "S"), path));
#endif

#if 0
  static struct option long_options[] =
  {
    {"scheme",   required_argument, 0, 's'},
    {"authority",     required_argument, 0, 'h'},
    {"dataprov", required_argument, 0, 'd'},
    {"open",     required_argument, 0, 'o'},
    {"cmd",      required_argument, 0, 'c'},
    {0, 0, 0, 0}
  };

  while (1)
  {
    int option_index = 0;

    int c = getopt_long(argc, argv, "s:h:d:o:c:", long_options, &option_index);

    if (c == -1) break;

    char p[256], *s;

    switch (c)
    {
      case 's':
        if (strcmp(optarg, "scdc+tcp") == 0) strncpy(scheme, "scdc+tcp", OPTARG_MAX);
        else if (strcmp(optarg, "scdc+uds") == 0) strncpy(scheme, "scdc+uds", OPTARG_MAX);
        else strncpy(scheme, "scdc", OPTARG_MAX);
        break;
      case 'a':
        strncpy(authority, optarg, OPTARG_MAX);
        break;
      case 'd':
        s = strchr(optarg, ':');
        s[0] = 0;
        ++s;
        sprintf(p, "%s%s", getenv("MERGE_SCDC_REPO_PATH"), s);
        dataprovs[ndataprovs++] = scdc_dataprov_open(optarg, p);
        break;
      case 'o':
        strncpy(open, optarg, OPTARG_MAX);
        break;
      case 'c':
        strncpy(cmdline, optarg, OPTARG_MAX);
        break;
      default:
        printf("unknown parameter: '%c'\n", c);
      }
  }

  if (strlen(scheme) > 0 || strlen(authority) > 0 || strlen(open) > 0)
  {
    char uri[1024];

    if (strlen(authority) > 0) sprintf(uri, "%s://%s/%s", scheme, authority, open);
    else sprintf(uri, "%s:/%s", scheme, open);

    printf("%suri: '%s'\n", prefix, uri);

    dataset = scdc_dataset_open(uri);

  } else dataset = SCDC_DATASET_NULL;

  printf("%sdataset: %p\n", prefix, dataset);

  run_cli(dataset, cmdline);

  scdc_dataset_close(dataset);
#endif

  while (ndataprovs > 0) scdc_dataprov_close(dataprovs[--ndataprovs]);

  scdc_release();

  printf("quit repository client\n");

  return 0;
}

File: test/repo.py/repo_cli.py
#!/usr/bin/python

#  
#  Copyright (C) 2014, 2015, 2016, 2017, 2018 Michael Hofmann
#  
#  This file is part of the Simulation Component and Data Coupling (SCDC) library.
#  
#  The SCDC library is free software: you can redistribute it and/or
#  modify it under the terms of the GNU Lesser Public License as published
#  by the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  The SCDC library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Lesser Public License for more details.
#  
#  You should have received a copy of the GNU Lesser Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#  


import os
import sys
import readline

import scdc
import repoH


#print("path: ", sys.path)

BASE_PATH = os.getenv('MERGE_SCDC_REPO_PATH', '')
MYSQL_CREDENTIALS = os.getenv('MERGE_SCDC_MYSQL_CREDENTIALS', '')

compX_WORKDIR = BASE_PATH + "demo/" + "compX/"


def manual_input_next(input):
  sys.stdout.write("INPUT #" + str(input.data) + " + <Enter>: ")
  b = sys.stdin.readline()
  if b == "":
    input.buf = None
    s = 0
  else:
    input.buf = b[:-1]
    s = len(input.buf)
  input.buf_size = s
  input.current_size = s
  input.total_size += input.current_size
  if s == 0:
    input.next = None
  input.data += 1
  return True

def manual_output_next(output):
  if output.format == "text" or output.format.split(":")[0] == "fslist":
    print "OUTPUT #" + str(output.data) + ": '" + output.buf2str() + "'"
  else:
    print "OUTPUT #" + str(output.data) + ": " + str(output.current_size) + " bytes"
  output.data = output.data + 1
  return True

def prepare_dataset_cmd(cmdline):
  print("cmdline: " + cmdline)
  input_pos = cmdline.find("<")
  output_pos = cmdline.find(">")

  cmdlen = len(cmdline)

  input_buf_size = 64
  output_buf_size = 256

  input_type = 0

  if input_pos >= 0:
    if input_pos < cmdlen:
      cmdlen = input_pos

    input_type = 1

    if output_pos > input_pos:
      input_str = cmdline[input_pos + 1:output_pos]
    else:
      input_str = cmdline[input_pos + 1:]
    print "input: '" + input_str + "'"

    name_pos = input_str.find(":")
    if name_pos >= 0:
      type_str = input_str[:name_pos]
      name_str = input_str[name_pos + 1:]
    else:
      type_str = None
      name_str = input_str

    if type_str == None:
      if os.path.isdir(name_str):
        type_str = "fs"
      elif os.path.isfile(name_str):
        type_str = "file"
      else:
        type_str = "text"

    if type_str == "fs":
      print "input fs: '" + name_str + "'"
      input = scdc.dataset_input_create("fs", name_str);
    elif type_str == "file":
      print "input file: '" + name_str + "'"
      input = scdc.dataset_input_create("file", name_str);
    elif type_str == "stream":
      print "input stream: '" + name_str + "'"
      input = scdc.dataset_input_create("stream", sys.stdin);
    elif type_str == "text":
      print "input text: '" + name_str + "'"
      input = scdc.dataset_input_create("buffer", name_str);
      input.format = "text"
    else:
      print "input manual:"
      input = scdc.dataset_input()
      input.format = "text"
      input.buf = "\x00" * input_buf_size
      input.buf_size = input_buf_size
      input.total_size = 0
      input.total_more = 1
      input.current_size = 0
      input.next = manual_input_next
      input.data = 0

  else:
    print "input none:"
    input = None


  output_type = 0

  if output_pos >= 0:
    if output_pos < cmdlen:
      cmdlen = output_pos

    output_type = 1

    if input_pos > output_pos:
      output_str = cmdline[output_pos + 1:input_pos]
    else:
      output_str = cmdline[output_pos + 1:]
    print "output: '" + output_str + "'"

    name_pos = output_str.find(":")
    if name_pos >= 0:
      type_str = output_str[:name_pos]
      name_str = output_str[name_pos + 1:]
    else:
      type_str = None
      name_str = output_str

    if type_str == None:
      type_str = "redirect"

    if type_str == "fs":
      print "output fs: '" + name_str + "'"
      output = scdc.dataset_output_create("fs", name_str);
    elif type_str == "file":
      print "output file: '" + name_str + "'"
      output = scdc.dataset_output_create("file", name_str);
    elif type_str == "stream":
      print "output stream: '" + name_str + "'"
      output = scdc.dataset_output_create("stream", sys.stdout);
    elif type_str == "consume":
      print "output consume: '" + name_str + "'"
      output = scdc.dataset_output_create("consume");
    elif type_str == "redirect":
      print "output redirect: '" + name_str + "'"
      output = scdc.dataset_output_create("alloc");
      output_type = 2
    else:
      print "output manual:"
      output = scdc.dataset_output()
      output.buf = "\x00" * output_buf_size
      output.buf_size = output_buf_size
      output.next = manual_output_next
      output.data = 0
      output_type = -1

  else:
    print "output loop manual:"
    output = scdc.dataset_output()
    output.buf = "\x00" * output_buf_size
    output.buf_size = output_buf_size

  return (cmdline[:cmdlen], input, input_type, output, output_type)


def unprepare_dataset_cmd(input, input_type, output, output_type):
  print "cmd input: " + str(input)

  if input_type > 0:
    scdc.dataset_input_destroy(input)
  else:
    pass

  print "cmd output: " + str(output)

  if output_type > 0:
    if output_type == 2:
      scdc.dataset_output_redirect(output, "stream", sys.stdout)
    scdc.dataset_output_destroy(output)
  else:
    if output_type == 0:
      i = 0
      while True:
        data = output.data
        output.data = i

        manual_output_next(output)

        i = output.data
        output.data = data

        if output.next is None:
          break

        if False:
          try:
            raw_input("Press <Enter>")
          except:
            pass

        output.next(output)


scdc.log_init("log_FILE", sys.stdout, sys.stderr)

scdc.init()

dataprovs = []

dataprovs.append(scdc.dataprov_open("repoA", "fs", BASE_PATH + "A"))
dataprovs.append(scdc.dataprov_open("repoB", "fs", BASE_PATH + "B"))
dataprovs.append(scdc.dataprov_open("repoC", "gen"))
#dataprovs.append(scdc.dataprov_open("repoD", "mysql", MYSQL_CREDENTIALS))
dataprovs.append(scdc.dataprov_open("repoH", "hook:id", repoH.repoH_hooks, 2501, 2502))
dataprovs.append(scdc.dataprov_open("repoJ", "jobrun", "uname -a; xterm -e sleep", BASE_PATH + "J"))
dataprovs.append(scdc.dataprov_open("repoJ1", "jobrun", "uname -a; echo \"J1\"; xterm -e sleep", BASE_PATH + "J"))
dataprovs.append(scdc.dataprov_open("repoJ2", "jobrun", "uname -a; echo \"J2\"; xterm -e sleep", BASE_PATH + "J"))
dataprovs.append(scdc.dataprov_open("repoJH", "jobrun:handler", repoH.repoH_jobrun_handler, True))
dataprovs.append(scdc.dataprov_open("repoRG", "register"))
dataprovs.append(scdc.dataprov_open("repoRL", "relay"))
dataprovs.append(scdc.dataprov_open("repoJRL", "jobrun_relay"))
dataprovs.append(scdc.dataprov_open("repoS", "fs_store", BASE_PATH + "S"))

#scdc.dataset_cmd(None, "scdc:///repoJRL/CONFIG/relay put sub scdc:///repoJ", None, None)
#scdc.dataset_cmd(None, "scdc:///repoJRL/CONFIG/relay put sub1 scdc:///repoJ1", None, None)
#scdc.dataset_cmd(None, "scdc:///repoJRL/CONFIG/relay put sub2 scdc:///repoJ2", None, None)

uri = None
cmdline = None

#cmdline = "scdc:/repoA/test_1k.in ls"

#cmdline = "\n\
#scdc:///repoJRL put test1 jobbegin;run 10;jobend\n\
#scdc:///repoJRL put test2 jobbegin;run 10;jobend\n\
#scdc:///repoJRL put test3 jobbegin;run 10;jobend\n\
#scdc:///repoJRL put test4 jobbegin;run 10;jobend\n\
#scdc:///repoJRL ls\n\
#"

#cmdline = "scdc:///repoJRL put test run 2\nscdc:///repoJRL ls\nscdc:///repoJRL rm test\nscdc:///repoJRL ls\nscdc:///repoJRL rm test"
#cmdline = "scdc:///repoJRL put test run 2\nscdc:///repoJRL/test rm\nscdc:///repoJRL ls"

#

#uri = "scdc:////CONFIG"
#uri = "scdc:///repoR"
#uri = "scdc+tcp://localhost/repoA"
#uri = "scdc+uds://repo_srv/repoA"
#uri = "scdc:/repoJ"

#cmdline = "put test_put_man.out<manual:"
#cmdline = "put test_put_text.out<text:ABCDEF"
#cmdline = "put test_put_file.out<file:test_2k.in"
#cmdline = "get test_2k.in"
#cmdline = "get test_2k.in>text:"
#cmdline = "get test_2k.in>file:test_put_file.out"

#cmdline = "ls"
#cmdline = "ls>file:TEST"
#cmdline = "get rand100M>file:TEST"
#cmdline = "put DEMO<file:TEST"
#cmdline = "get in2out<text:TEST"
#cmdline = "get in4out<text:TEST"
#cmdline = "get in5out<text:TEST"

#uri = "scdc:/repoC/ascii/zero"
#cmdline = "put TEST<file:test.in"
#cmdline = "get 70"
#cmdline = "get 70>probe:"

#uri = "scdc:///repoA/CONFIG/root"
#cmdline = "cd type\ninfo\ncd\nls"

#uri = "scdc:/repoH/"
#uri = "scdc+tcp://localhost/repoH/"
#cmdline = "ls"
#cmdline = "put test_put_man.out"
#cmdline = "put test_put_text.out<text:ABCDEF"
#cmdline = "put test_put_file.out<file:test_2k.in"
#cmdline = "get test_get"
#cmdline = "get test_get>text:"
#cmdline = "get test_get>file:test_get_file.out"

#cmdline = "scdc:/repoR put scdc:///repoA\nscdc:/repoR put scdc:///repoB\nscdc:/repoR info scdc:///repoA"
#cmdline = "scdc:/repoR put scdc:///repoA\nscdc:/repoR info scdc:///repoA"
#cmdline = "scdc:///repoH/CONFIG ls"

#cmdline = "get param2,*"

#cmdline = "scdc:/repoJ ls"

#cmdline = "scdc+uds://repo_srv/repoH get x20"
#cmdline = "scdc+uds://repo_srv/repoA ls\nscdc+uds://repo_srv/repoB ls"
#cmdline = "scdc:///repoA ls\nscdc:///repoB ls"

#cmdline = "scdc+tcp://localhost/repoA ls\nscdc+tcp:///repoB ls"

#cmdline = "scdc:///repoJ/CONFIG get type,max_parallel_jobs"
#cmdline = "scdc:///repoRG/CONFIG ls"
#cmdline = "scdc:///repoRG/CONFIG get type"
#cmdline = "scdc:///repoRG/CONFIG/type get"

#cmdline = "scdc:///repoJRL/CONFIG/relay put sub scdc:///repoA\nscdc:///repoJRL/CONFIG/relay put sub scdc:///repoJ\nscdc:///repoJRL/CONFIG/relay ls"

#cmdline = "scdc:///repoJRL/CONFIG/relay put sub scdc:///repoJ"

#cmdline = "scdc:///repoJ/CONFIG put nodes nodeA:8:1.5\n\
#scdc:///repoJ/CONFIG put nodes nodeB:4:2.5\n\
#scdc:///repoJ/CONFIG/nodes get\n\
#scdc:///repoJ/CONFIG/runjobs get"
#cmdline = "scdc:///repoJRL/CONFIG get relay"

uri = "scdc:///repoA"
#cmdline = "ls rand10M\ncd rand10M\nls\ncd\nls xxxxx"
#cmdline = "ls"


if uri:
  dataset = scdc.dataset_open(uri)
else:
  dataset = None
print("dataset: " + str(dataset))

if cmdline == False: cmdline = "quit"

if isinstance(cmdline, str):
  cmdlines = cmdline.split("\n")
  cmdlines_max = len(cmdlines)
else:
  cmdlines = None
  cmdlines_max = -1

n = 0
#while (dataset and cmdlines_max < 0) or (dataset != False and n < cmdlines_max):
while (cmdlines_max < 0) or (dataset != False and n < cmdlines_max):

  if cmdlines == None:
    try:
      c = raw_input("Enter command for dataset '" + str(dataset) + "': ")
    except:
      c = "quit"
  else:
    c = cmdlines[n]
    n += 1

  cs = c.split()

  if len(cs) == 0:
    continue

  if len(cs) == 1 and cs[0] == "quit":
    break

  if len(cs) == 2 and cs[0] == "open":
    if dataset: scdc.dataset_close(dataset)
    dataset = scdc.dataset_open(cs[1])
    continue

  if len(cs) == 1 and cs[0] == "close":
    scdc.dataset_close(dataset)
    dataset = None
    continue

  (cmd, input, input_type, output, output_type) = prepare_dataset_cmd(c)

  print("cmd: '" + cmd + "'");
  res = scdc.dataset_cmd(dataset, cmd, input, output)
  print("result: " + str(res));

  unprepare_dataset_cmd(input, input_type, output, output_type)

scdc.dataset_close(dataset)

while len(dataprovs) > 0:
  dp = dataprovs.pop()
  scdc.dataprov_close(dp)

scdc.release()

scdc.log_release()
